### PR TITLE
blocking sync in view changer

### PR DESCRIPTION
**Pull request by HagarMeir**
Signed-off-by: Hagar Meir <hagar.meir@ibm.com>

### CODE REVIEW

File: internal/bft/viewchanger.go [hunk 1]
```diff
@@ -607,14 +616,30 @@ func (v *ViewChanger) commitLastDecision(lastDecisionSequence uint64, lastDecisi
...
 	}
 	if md.LatestSequence > lastDecisionSequence+1 {
 		v.Logger.Panicf("Node %d has a checkpoint for sequence %d which is much greater than the last decision sequence %d", v.SelfID, md.LatestSequence, lastDecisionSequence)
 	}
 }
 
+func (v *ViewChanger) sync(targetSeq uint64) {
+	for {
+		v.Synchronizer.Sync()
+		select { // wait for sync to return with expected info
```
Comment (yacovm) [inline internal/bft/viewchanger.go:630]: I think you should also wait for startChangeChan for completeness, just to un-block anyone that is invoking StartViewChange.

**Review by yacovm — CHANGES_REQUESTED** [review]
Please see comment in the code + maybe we should add a test in the **view_changer_test.go** for this,

**Comment by HagarMeir** [comment]
I can add unit tests on Sunday

File: internal/bft/viewchanger_test.go [hunk 2]
```diff
@@ -727,6 +727,121 @@ func TestCommitLastDecision(t *testing.T) {
...
+		Content: &protos.Message_NewView{
+			NewView: &protos.NewView{
+				SignedViewData: signed,
+			},
+		},
+	}
+
+	vc.Start(1)
+
+	synchronizerWG.Add(2)
```
Comment (yacovm) [inline internal/bft/viewchanger_test.go:831]: why is it safe to add the counter when the view changer operates in a different goroutine?

File: internal/bft/viewchanger_test.go [hunk 3]
```diff
@@ -727,6 +727,121 @@ func TestCommitLastDecision(t *testing.T) {
...
+
+	vc.HandleMessage(1, msg)
```
Comment (yacovm) [inline internal/bft/viewchanger_test.go:835]: can you perhaps add some comments to give background what is the scenario, what is the expected behavior, etc?

File: internal/bft/viewchanger.go [hunk 4]
```diff
@@ -607,14 +616,33 @@ func (v *ViewChanger) commitLastDecision(lastDecisionSequence uint64, lastDecisi
...
+		case info := <-v.informChan:
```
Comment (yacovm) [inline internal/bft/viewchanger.go:631]: if you read from **informChan** don't you also need to call **v.informNewView(info)** ?

Comment (HagarMeir) [reply internal/bft/viewchanger.go:631]: this is why I added a TODO later
`// TODO what if view is bigger than expected?`
it won't be simply calling the `informNewView`
and I will deal with this in a later PR

**Review by yacovm — APPROVED** [review]

**Pull request merged**